proxy/src/control/cache.rs

```diff
@@ -54,37 +63,47 @@ impl<T> Cache<T> where T: Clone + Copy + Eq + std::hash::Hash {
     }
 
     pub fn extend<I, F>(&mut self, iter: I, on_change: &mut F)
-        where I: Iterator<Item = T>,
-              F: FnMut(T, CacheChange),
+    where
+        I: Iterator<Item = (K, V)>,
+        F: FnMut((K, V), CacheChange),
     {
-        fn extend_inner<T, I, F>(values: &mut HashSet<T>, iter: I, on_change: &mut F)
-            where T: Copy + Eq + std::hash::Hash, I: Iterator<Item = T>, F: FnMut(T, CacheChange)
+        fn extend_inner<K, V, I, F>(values: &mut IndexMap<K, V>, iter: I, on_change: &mut F)
+        where
+            K: Eq + Hash + Copy + Clone,
+            V: Clone,
+            I: Iterator<Item = (K, V)>,
+            F: FnMut((K, V), CacheChange),
         {
-            for value in iter {
-                if values.insert(value) {
-                    on_change(value, CacheChange::Insertion);
+            for (key, value) in iter {
+                if values.insert(key, value.clone()).is_none() {
```

Comment:
In the event that the value already existed, we need to compare the labels. If the labels are different between the new and old, then we need to call `on_change(..., CacheChange::Replace)`. If the labels are the same then we don't call `on_change()`.

Reply:
And, if there's ever any ambiguity, we should default to overpublishing, rather than underpublishing (i.e. at-least-once not at-most-once).